Tensile/KernelWriterAssembly.py

```diff
@@ -9155,6 +9192,274 @@ def notLocalFullTileElements(self, kernel, edge):
         return self.notLocalFullTileElementsMFMA(kernel, edge)
     else:
         return self.notLocalFullTileElementsVALU(kernel, edge)
+  ##############################################################################
+  # Store Remap: Local Write
+  ##############################################################################
+  def storeRemapAddLocalWrite(self, kernel, ss, addrCalc, sumIdx):
+    """
+    Add localWrite for the element with addrCalc and sumIdx.
+    """
+    kStr = ""
+
+    bps = kernel["ProblemType"]["DataType"].numBytes() * ss.cfg.gwvw
+    rpv = kernel["ProblemType"]["DataType"].numRegisters() * ss.cfg.gwvw
+
+    addr0 = vgpr(self.storeRemapLW)
+    offset =  addrCalc.globalOffset
+
+    # only consider store with vector4
+    # bps==8 is for bf16 and fp16
+    # bps==16 is for fp32
+    if bps==8:
+      kStr += inst("ds_write_b64", addr0, vgpr(sumIdx, rpv), \
+                 "offset:%u"%offset, "storeRemap lw")
+    elif bps==16:
+      kStr += inst("ds_write_b128", addr0, vgpr(sumIdx, rpv), \
+                 "offset:%u"%offset, "storeRemap lw")
+    else:
+       assert ("StoreRemap: bad bps! only support vector4 local write")
+
+    return kStr
+
+  ##############################################################################
+  # Store Remap: Local Read and Global Write
+  ##############################################################################
+  def storeRemapAddStore(self, kernel, ss, addrCalc, tmpVgpr, tmpS01, edge):
+    kStr = ""
+
+    kStr += inst("s_waitcnt", "lgkmcnt(0)", "wait for LDS write" )
+    kStr += "\n"
+
+    gwvw = kernel["StoreRemapVectorWidth"]
+    startIdx = self.storeRemapStartSumIdx
+    endIdx = self.storeRemapEndSumIdx
+    bpe = kernel["ProblemType"]["DataType"].numBytes()
+    bps = kernel["ProblemType"]["DataType"].numBytes() * gwvw
+    rpe = kernel["ProblemType"]["DataType"].numRegisters()
+    rpv = kernel["ProblemType"]["DataType"].numRegisters() * gwvw
+
+    src = vgpr(self.storeRemapLR)
+    for i in range (startIdx, endIdx, gwvw):
```

Comment:
i wonder b2b ds_read_xx affect inversely for mi32x32 with bigger MT0?  given that there isnt much math left to do to hide wrte->read-> global-write..

Reply:
can you think of using existing store code and add waitcnt() when needed?. may be for next optimization..

Reply:
> i wonder b2b ds_read_xx affect inversely for mi32x32 with bigger MT0? given that there isnt much math left to do to hide wrte->read-> global-write..

It wouldn't really happen due to LDS resource has limitation.
For example: 
fp16 MT256x256 with mi32x32 and 4 waves, it would consume (256(MT0)+4 padding) x 32 (MI32x32)x 4 (waves) * 2 (bpe) = 66560.bytes lds.
This kernel will be rejected by solutionStruction.py

Reply:
> can you think of using existing store code and add waitcnt() when needed?. may be for next optimization..

I ever did this in my implementation.
But I think that the code is better separated cleanly with existing store code.
For storeRemap store code, it needs to consider "local read" and "edge" and "waitcnt".
It would be more complicated and original store code is very simple.
Separate those would be more easily for reading and maintenance.

Thanks for your suggestions anyway.